### PR TITLE
feat(n8n Form Page Node): New node

### PR DESCRIPTION
**Pull request by michael-radency**
## Summary

New **n8n Form** that allows the creation of complex multi-step forms and is fully integrated into the workflow, enabling the use of incoming data and branching logic.

**n8n Form** requires the **n8n Form Trigger** to be set as the workflow's trigger.

- Select the type of page: form page or completion screen.
- Customize each form step's title, description, and button label, or inherit these settings from the trigger.
- Define form fields using parameters or send them as JSON.
- Customize the URL with query parameters, which can be used for field defaults. These parameters are also available in the n8n Form nodes connected to that trigger.
image
image
image
image

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/NODE-1474/spike-multi-step-form
https://linear.app/n8n/issue/NODE-1472/add-form-node-p0
https://community.n8n.io/t/how-to-use-data-in-formtrigger-node/56897

### CODE REVIEW

**Comment by J-Fo-S** [comment]
hi, when's this going to merge?

**Comment by michael-radency** [comment]
> hi @michael-radency, do you have any expectation of when this will merge?

Hello @J-Fo-S this PR still in development, approximately this month

File: packages/editor-ui/src/composables/useRunWorkflow.ts [hunk 1]
```diff
@@ -390,24 +399,39 @@ export function useRunWorkflow(useRunWorkflowOpts: { router: ReturnType<typeof u
...
+									const formTrigger = execution?.workflowData.nodes.find((node) => {
+										return node.type === FORM_TRIGGER_NODE_TYPE;
+									});
+									const runNodeFilter = execution?.data?.startData?.runNodeFilter || [];
+									if (formTrigger && !runNodeFilter.includes(formTrigger.name)) {
+										isFormShown = true;
+									}
+								}
+								if (!isFormShown) {
+									let testUrl = '';
```
Comment (ShireenMissi) [inline packages/editor-ui/src/composables/useRunWorkflow.ts:425]:
```suggestion						
```

File: packages/editor-ui/src/composables/useRunWorkflow.ts [hunk 2]
```diff
@@ -390,24 +399,39 @@ export function useRunWorkflow(useRunWorkflowOpts: { router: ReturnType<typeof u
...
 						if (
-							waitingNode &&
-							waitingNode.type === WAIT_NODE_TYPE &&
-							waitingNode.parameters.resume === 'form'
+							lastNode &&
+							(lastNode.type === FORM_NODE_TYPE ||
+								(lastNode.type === WAIT_NODE_TYPE && lastNode.parameters.resume === 'form'))
 						) {
-							const testUrl = getFormResumeUrl(waitingNode, executionId as string);
+							const testUrl = getFormResumeUrl(lastNode, executionId as string);
```
Comment (ShireenMissi) [inline packages/editor-ui/src/composables/useRunWorkflow.ts:407]:
```suggestion
							let testUrl = getFormResumeUrl(lastNode, executionId as string);
```

File: packages/workflow/src/TypeValidation.ts [hunk 3]
```diff
@@ -148,6 +153,96 @@ export const tryToParseObject = (value: unknown): object => {
...
+	'dropdown',
+	'email',
+	'file',
+	'number',
+	'password',
+	'text',
+	'textarea',
+];
+
+export const tryToParseFormFields = (value: unknown): FormFieldsParameter => {
```
Comment (ShireenMissi) [inline packages/workflow/src/TypeValidation.ts:179]: I suggest including json in the name to make it clear what we are parsing from
```suggestion
export const tryToParseJsonToFormFields = (value: unknown): FormFieldsParameter => {
```

File: packages/nodes-base/nodes/Form/Form.node.ts [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+const completionProperties = updateDisplayOptions(
+	{
+		show: {
+			operation: ['completion'],
+		},
+	},
+	[
+		{
+			displayName: 'On N8n Form Submission',
```
Comment (ShireenMissi) [inline packages/nodes-base/nodes/Form/Form.node.ts:95]:
```suggestion
			displayName: 'On n8n Form Submission',
```

**Review by ShireenMissi — COMMENTED** [review]
Mostly looks good 🎉 left some minor comments
also please fix the failing frontend test 🙏

**Review by ShireenMissi — DISMISSED** [review]
LGTM 🎉

**Review by ShireenMissi — APPROVED** [review]
LGTM 🎉

**Comment by janober** [comment]
Got released with [`n8n@1.65.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.65.0)

**Comment by mickaelandrieu** [comment]
Sorry to interact here, but you are probably the best people to answer me :
I'd like to introduce the markdown support of the Form Description.
By any chance, can someone help me with that (as an open source contribution) and is there any chance to be merged ?

**Comment by ashleydoidge** [comment]
Hi @janober @michael-radency, is there a timeout included as part of the form trigger that must be adhered to?

There are a few more transformation steps that I need to do in order to build the second page of my form, using dynamic values to set up some questions & answers, all in all this takes around 5 seconds between the form submission and when the second form should pop up, however the second form never pops up/overwrites the first form, so I have to trigger it manually in my workflow (Which of course wouldn't work in any real life scenarios)

If I re-arrange the form so it's right next to any built in nodes and doesn't include my API calls, it pops up fine as the transformation nodes are instant, but it seems as if the window to trigger the second form is very small and does not allow for any further transformations that I'm using with some API calls to our CRM that extends the flow time to a few seconds.

Any help would be appreciated.

**Comment by J-Fo-S** [comment]
Hi @ashleydoidge you'll probably get a quicker/better response if you ask this in n8n community forum

**Comment by ashleydoidge** [comment]
> Hi @ashleydoidge you'll probably get a quicker/better response if you ask this in n8n community forum

Thanks, also posted there now.

**Pull request merged**